// mysqlpool/pool.js
var mysql = require('mysql');

module.exports = (function(){
	var _connections = [], i, c, current = 0;
	function Pool(opts) {
		if(typeof opts === "string"){
			// we only got the connection string
			opts = {url: opts}
		}
		opts = {
			url: opts.url,
			connections: opts.connections || 50,
			reconnect: opts.reconnect || true
		};
		if(opts.connections < 1)
			throw new Error('Connections must be more than one');
		i = 0;
		while(i < opts.connections) {
			c = new mysql.createConnection(opts.url);
			c.connect();
			_connections.push(c);
			handleDc(i);
			i++;
		}
		function handleDc(i){
			_connections[i].on('error', function(err) {
				if(!err.fatal)
					return;
				if(err.code !== 'PROTOCOL_CONNECTION_LOST')
					throw err;

				// console.log('Re-connecting lost connection: ' + err.stack);
				_connections[i] = new mysql.createConnection(_connections[i].config);
				_connections[i].connect();
				handleDc(i);
			});
		}
	}

	Pool.prototype = {
		getClient: function(){
			if(_connections.length >= current)
				current = 0;
			return _connections[current++];
		},
		query: function(){
			this.getClient().query(arguments);
		}
	}

	return Pool;
})();
